rfcs/0014-opendal.md

```diff
@@ -0,0 +1,196 @@
+### Constraints
+
+Given that `object_store` is a core dependency deeply integrated throughout SlateDB, we must ensure:
+
+1. **No breaking changes**: The migration must be transparent to SlateDB users. Since we currently expose the `object_store::ObjectStore` trait as our public I/O API, we need to maintain API compatibility—either by keeping the same trait interface with a compatibility layer that preserves existing behavior.
+2. **Performance parity**: I/O performance must not regress. Critical paths (read, write, list operations) should be benchmarked to ensure OpenDAL meets or exceeds current performance metrics.
+3. **Feature completeness**: All existing functionality (caching, retrying, error handling) must be preserved or improved during migration.
+4. **Validation before rollout**: A PoC must be implemented and validated against our test suite before committing to the full migration. The transition should be incremental and allow rollback if issues arise.
+
+<!-- TOC --><a name="scope"></a>
+## Scope
+
+This RFC aims to accomplish the following:
+
+1. **Assess migration effort**: Document our current usage of `object_store` throughout the codebase and evaluate the feasibility of migrating to OpenDAL equivalents
+2. **Design migration path**: Develop a concrete, step-by-step migration strategy that satisfies the constraints outlined above
+3. **Validate OpenDAL compatibility**: Verify that OpenDAL's features and layers can pass our existing test suite and meet our functional requirements
+
+<!-- TOC --><a name="proposal"></a>
+## Proposal
+
+<!-- TOC --><a name="migration-strategy"></a>
+### Migration Strategy
+
+OpenDAL provides a **compatibility layer** that accepts `object_store::ObjectStore` implementations as backends. This allows us to maintain API compatibility with existing users while internally leveraging OpenDAL's layer system.
+
+The migration approach is straightforward: we keep the public API unchanged while internally wrapping user-provided object stores with OpenDAL's compatibility layer. This wrapper enables us to compose OpenDAL's layers (retry, metrics, caching) on top of any `object_store` implementation:
+
+```rust
+let store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let operator = Operator::new(ObjectStoreBuilder::new(store).build()?)
+    .layer(RetryLayer::new())
+    .layer(MetricsLayer::new())
+    .layer(cache_layer)
+    .finish()?;
+```
+
+This compatibility layer enables a low-risk, incremental migration strategy. In the initial phase, complex wrappers like `CachedObjectStore` can remain unchanged and be passed directly to OpenDAL's compatibility layer. This ensures zero behavioral changes for systems that heavily depend on current caching semantics, such as ZeroFS.
+
+Over time, we can evaluate OpenDAL's built-in layers for feature parity, replace custom wrappers with OpenDAL equivalents where beneficial, and contribute missing functionality upstream to OpenDAL if needed.
```

Comment:
I prefer we evaluate OpenDAL built-in layers for feature parity _before_ making changes. I worry we might do this migration then discover a bunch of issues with OpenDAL. I don't want to add OpenDAL stuff unless we're pretty confident it's reasonably doable and provides a significant net positive. If all we end up with is our custom logic stuffed into an OpenDAL backend, I would consider that a net loss.

I don't think we need to be completely exhaustive, but I would like to feel pretty confident we know what needs to be done and we know we're not going to take a big perf hit.

Reply:
What's the difference between CacheLayer and FoyerLayer? 🤔 forgive my ignorance

Reply:
> the main concern is mostly the progress of the cache layer from the OpenDal community

Yea, I don't want to migrate until OpenDAL has a cache layer we can use.

Reply:
> What's the difference between CacheLayer and FoyerLayer? 🤔 forgive my ignorance

iiuc, @Xuanwo plans to add a concrete implementation first, before generalizing it. for example, `FoyerLayer` would be a layer that hard-codes the dependency on `foyer` for caching, whereas `CacheLayer` will likely become a generic cache layer that can be plugged into different backends.

(however i kinda skeptical about whether there'll be any better cache backend exists than `foyer` 😲...

Reply:
> however i kinda skeptical about whether there'll be any better cache backend exists than foyer

Same! 😆  So starting with `FoyerLayer` is good enough for me. :)